File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/nurikabe/data/NurikabeDocument.java
package com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.data;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocument;
import com.zwstudio.logicpuzzlesandroid.common.data.GameProgress;
import com.zwstudio.logicpuzzlesandroid.common.data.MoveProgress;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.domain.NurikabeGame;
import com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.domain.NurikabeGameMove;
import com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.domain.NurikabeObject;

import org.androidannotations.annotations.EBean;

/**
 * Created by zwvista on 2016/09/29.
 */

@EBean
public class NurikabeDocument extends GameDocument<NurikabeGame, NurikabeGameMove> {
    protected void saveMove(NurikabeGameMove move, MoveProgress rec) {
        rec.row = move.p.row;
        rec.col = move.p.col;
        rec.strValue1 = move.obj.objAsString();
    }
    public NurikabeGameMove loadMove(MoveProgress rec) {
        return new NurikabeGameMove() {{
            p = new Position(rec.row, rec.col);
            obj = NurikabeObject.objTypeFromString(rec.strValue1);
        }};
    }
    public int getMarkerOption() {
        String o = gameProgress().option1;
        return o == null ? 0 : Integer.parseInt(o);
    }
    public void setMarkerOption(GameProgress rec, int o) {
        rec.option1 = String.valueOf(o);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/nurikabe/data/NurikabeDocument.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.data

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocument
import com.zwstudio.logicpuzzlesandroid.common.data.MoveProgress
import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.domain.NurikabeGame
import com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.domain.NurikabeGameMove
import com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.domain.NurikabeObject
import org.androidannotations.annotations.EBean

@EBean
class NurikabeDocument : GameDocument<NurikabeGame, NurikabeGameMove>() {
    override fun saveMove(move: NurikabeGameMove, rec: MoveProgress) {
        rec.row = move.p.row
        rec.col = move.p.col
        rec.strValue1 = move.obj.objAsString()
    }

    override fun loadMove(rec: MoveProgress) =
        NurikabeGameMove(Position(rec.row, rec.col), NurikabeObject.objTypeFromString(rec.strValue1!!))
}